### DISCUSSION TITLE
Pulsar IO threads vs pulsar-client-io-#-#

### DISCUSSION THREAD

**Discussion by rishavkg** [apache/pulsar#21459] [Q&A]
Hi,
I am new to apache pulsar .
I wanted to understand what is/are the purpose IO threads defined while creating pulsar client using pulsar client java library.
And what are the pulsar-client-io-#-# threads that can be seen in the logs. Are these connections in the connection pool and these are always greater than the io thread.

Is there any configuration defined in the pulsar client to limit the size of the connection pool if there is any connection pool?

**Comment by jak78** [comment]
Hi @rishavkg! Welcome to the Pulsar community!

IO threads in the Pulsar Client are responsible for handling the I/O operations related to network communication between the client and the brokers. 
The threads named pulsar-client-io-#-# that you see in the logs are indeed the IO threads. Each one is responsible for handling I/O operations, as mentioned above. 
Apache Pulsar does maintain a connection pool to efficiently manage and reuse connections to brokers. The connection pool size isn't directly configurable by the number, but indirectly, by tuning the number of IO threads and the max number of connections per broker, you can influence the number of active connections.

For example, to set the number of IO threads:
`clientBuilder.ioThreads(numberOfIOThreads); // Defaults to the number of available processors
`
And to set the max number of connections per broker:
`clientBuilder.connectionsPerBroker(maxConnections); // Defaults to one. It should be increased only if you need to improve producer throughput over a high-latency network.
`

While it's tempting to adjust these numbers, it's essential to understand your workload's requirements. Setting them too low could throttle your client's throughput, while setting them unnecessarily high could lead to resource wastage. If you're just starting, it's usually good to stick with the defaults and only adjust if you notice performance issues or have a clear understanding of your workload's needs.

Hope this helps,

**Reply by rishavkg** [reply]
Hi @jak78 , Thanks for providing clarity.
Although we are not setting any IOThread. We are using default configuration with a single pulsar client to create all consumers and producers.
We can see thread names up to pulsar-client-1-129.
Does these number signifies anything? I mean if there are only  8 cores then why its going up to 129.

**Reply by jak78** [reply]
Seeing a thread number like 129 doesn’t mean there are 129 active IO threads. Some threads might have been shut down, and new ones created, but the naming convention will keep incrementing.
